src/utils/generate_cxx_code.cpp

```diff
@@ -170,6 +170,297 @@ generate_cxx_code::get_all_dependencies(
+}
+
+
+/**
+ * Visits the given ASTNode as the function "log(10, x)" and in doing so,
+ * formats it as "log10(x)" (where x is any subexpression).
+ */
+void
+generate_cxx_code::FormulaFormatter_visitLog10_wcs ( const ASTNode_t *parent,
+                              const ASTNode_t *node,
+                              StringBuffer_t  *sb )
+{
+  StringBuffer_append(sb, "log10(");
+  generate_cxx_code::FormulaFormatter_visit_wcs(node, ASTNode_getChild(node, 1), sb);
+  StringBuffer_appendChar(sb, ')');
+}
+
+
+/**
+ * Visits the given ASTNode as the function "root(2, x)" and in doing so,
+ * formats it as "sqrt(x)" (where x is any subexpression).
+ */
+void
+generate_cxx_code::FormulaFormatter_visitSqrt_wcs ( const ASTNode_t *parent,
+                             const ASTNode_t *node,
+                             StringBuffer_t  *sb )
+{
+  StringBuffer_append(sb, "sqrt(");
+  generate_cxx_code::FormulaFormatter_visit_wcs(node, ASTNode_getChild(node, 1), sb);
+  StringBuffer_appendChar(sb, ')');
+}
+
+
+/**
+ * Visits the given ASTNode as a unary minus.  For this node only the
+ * traversal is preorder.
+ */
+void
+generate_cxx_code::FormulaFormatter_visitUMinus_wcs ( const ASTNode_t *parent,
+                               const ASTNode_t *node,
+                               StringBuffer_t  *sb )
+{
+  StringBuffer_appendChar(sb, '-');
+  generate_cxx_code::FormulaFormatter_visit_wcs ( node, ASTNode_getLeftChild(node), sb );
+}
+
+
+/**
+ * Visits the given ASTNode and continues the inorder traversal.
+ */
+void
+generate_cxx_code::FormulaFormatter_visitOther_wcs ( const ASTNode_t *parent,
+                              const ASTNode_t *node,
+                              StringBuffer_t  *sb )
+{
+  unsigned int numChildren = ASTNode_getNumChildren(node);
+  int group       = FormulaFormatter_isGrouped(parent, node);
+  unsigned int n;
+
+
+  if (group)
+  {
+    StringBuffer_appendChar(sb, '(');
+  }
+
+  if (numChildren == 0) {
+    FormulaFormatter_format(sb, node);
+  }
+
+  else if (numChildren == 1)
+  {
+    //I believe this would only be called for invalid ASTNode setups,
+    // but this could in theory occur.  This is the safest 
+    // behavior I can think of.
+    FormulaFormatter_format(sb, node);
+    StringBuffer_appendChar(sb, '(');
+    generate_cxx_code::FormulaFormatter_visit_wcs( node, ASTNode_getChild(node, 0), sb );
+    StringBuffer_appendChar(sb, ')');
+  }
+
+  else {
+    generate_cxx_code::FormulaFormatter_visit_wcs( node, ASTNode_getChild(node, 0), sb );
+
+    for (n = 1; n < numChildren; n++)
+    {
+      FormulaFormatter_format(sb, node);
+      generate_cxx_code::FormulaFormatter_visit_wcs( node, ASTNode_getChild(node, n), sb );
+    }
+  }
+
+  if (group)
+  {
+    StringBuffer_appendChar(sb, ')');
+  }
+}
+
+
+
+/**
+ * Visits the given ASTNode node.  This function is really just a
+ * dispatcher to either SBML_formulaToString_visitFunction() or
+ * SBML_formulaToString_visitOther().
```

Comment:
Is this comment from SBML source? If so, explain your modification here.

Reply:
I put a comment line for explaining our modifications.

Reply:
I put all the functions borrowed from SBMl into a separate file.